include/dmlc/optional.h

```diff
@@ -31,40 +31,148 @@ struct nullopt_t {
 
 /*! Assign null to optional: optional<T> x = nullopt; */
 constexpr const nullopt_t nullopt = nullopt_t(0);
+/*! \brief C++14 aliases */
+template <typename T> using decay_t = typename std::decay<T>::type;
+template <bool B, typename T = void>
+using enable_if_t = typename std::enable_if<B, T>::type;
 
+/*! \brief disambiguation tags that can be passed to the constructors of
+ * dmlc::optional. A tag type to tell constructor to construct its value in-place
+ */
+struct in_place_t {
+  explicit in_place_t() = default;  // NOLINT(*)
+};
+/*! \brief A tag to tell constructor to construct its value in-place */
+static constexpr in_place_t in_place{};
+
+/*! \brief  Is not constructible or convertible from any expression of type
+(possibly const) dmlc::optional<U>, i.e., the following 8 type traits are all
+false: Link: https://en.cppreference.com/w/cpp/utility/optional/optional */
+template <typename T> class optional;
+template <typename T, typename U, typename Other>
+using enable_constructor_from_other =
+    enable_if_t<std::is_constructible<T, Other>::value &&
+                !std::is_constructible<T, optional<U> &>::value &&
+                !std::is_constructible<T, optional<U> &&>::value &&
+                !std::is_convertible<optional<U> &, T>::value &&
+                !std::is_convertible<optional<U> &&, T>::value &&
+                !std::is_constructible<T, const optional<U> &>::value &&
+                !std::is_constructible<T, const optional<U> &&>::value &&
+                !std::is_convertible<const optional<U> &, T>::value &&
+                !std::is_convertible<const optional<U> &&, T>::value>;
+template <typename T, typename U>
+using enable_constructor_from_value =
+    enable_if_t<std::is_constructible<T, U &&>::value &&
+                !std::is_same<decay_t<U>, in_place_t>::value &&
+                !std::is_same<optional<T>, decay_t<U>>::value>;
 /*!
  * \brief c++17 compatible optional class.
  *
  * At any time an optional<T> instance either
  * hold no value (string representation "None")
  * or hold a value of type T.
  */
-template<typename T>
+template <typename T>
 class optional {
  public:
-  /*! \brief construct an optional object that contains no value */
+  using value_type = T;
+  /*! \brief constructs an object that does not contain a value. */
   optional() : is_none(true) {}
-  /*! \brief construct an optional object with value */
-  explicit optional(const T& value) {
-#pragma GCC diagnostic push
-#if __GNUC__ >= 6
-#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
-#endif
+  /*! \brief constructs an object that does contain a nullopt value. */
+  optional(dmlc::nullopt_t) noexcept : is_none(true) {} // NOLINT(*)
+  /*! \brief copy constructor, if other contains a value, then stored value is
+   * direct-intialized with it. */
+  optional(const optional &other) = default;
```

Comment:
it seems that reverting the use of the default copy constructor here stops the segfault